Readme.md
Introduction
============
Topic based course format with the ability to arrange the topics in columns except 0.

If you find an issue with the format, please see the 'Reporting Issues' section below.

[![Build Status](https://travis-ci.org/gjb2048/moodle-format_columns.svg?branch=master)](https://travis-ci.org/gjb2048/moodle-format_columns)

Required version of Moodle
==========================
This version works with Moodle version 2015111600.00 3.0 (Build: 20151116) and above within the 3.0 branch until the
next release.

Please ensure that your hardware and software complies with 'Requirements' in 'Installing Moodle' on
'docs.moodle.org/30/en/Installing_Moodle'.

Supporting Columns development
==============================
If you find Columns useful and beneficial, please consider sponsoring by:

PayPal - Please contact me via my 'Moodle profile' (above) for details as I am an individual and therefore am unable
to have 'buy me now' buttons under their terms.

Flattr - https://flattr.com/profile/gjb2048

I develop and maintain for free and any sponsorships to assist me in this endeavour are appreciated.

Installation
============
 1. Ensure you have the version of Moodle as stated above in 'Required version of Moodle'.  This is essential as the
    format relies on underlying core code that is out of my control.
 2. If upgrading from Moodle 2.3 or a previous release, please see 'Upgrade Instructions' below.
 3. Put Moodle in 'Maintenance Mode' (docs.moodle.org/en/admin/setting/maintenancemode) so that there are no 
    users using it bar you as the administrator - if you have not already done so.
 4. Copy 'columns' to '/course/format/' if you have not already done so.
 5. In 'cnconfig.php' change the value of defaultcolumns' for setting the default layout, structure and columns respectively for
    new / updating courses as desired by following the instructions contained within.
 6. Login as an administrator and follow standard the 'plugin' update notification.  If needed, go to
    'Site administration' -> 'Notifications' if this does not happen.
 7. Put Moodle out of Maintenance Mode.

Upgrade Instructions
====================
1. Ensure you have the version of Moodle as stated above in 'Required version of Moodle'.  This is essential as the
   format relies on underlying core code that is out of my control.
3. Put Moodle in 'Maintenance Mode' so that there are no users using it bar you as the administrator.
4. In '/course/format/' move old 'columns' directory to a backup folder outside of Moodle.
5. Follow installation instructions above.
6. Perform a 'Purge all caches' under 'Home -> Site administration -> Development -> Purge all caches'.
7. Put Moodle out of Maintenance Mode.

Uninstallation
==============
1. Put Moodle in 'Maintenance Mode' so that there are no users using it bar you as the administrator.
2. It is recommended but not essential to change all of the courses that use the format to another.  If this is
   not done Moodle will pick the last format in your list of formats to use but display in 'Edit settings' of the
   course the first format in the list.  You can then set the desired format.
3. In '/course/format/' remove the folder 'columns'.
4. In the database, remove the entry for 'format_columns' ('plugin' attribute) in the table 'config_plugins'.  
   If using the default prefix this will be 'mdl_config_plugins'.
5. Put Moodle out of Maintenance Mode.

Course Backup and Restore Instructions
======================================
1. Backup as you would any other course.  The number of columns will be stored with the course settings.
2. Restore as you would any other course.  If you are offered the option of 'Overwrite Course Configuration'
   you must say 'Yes' to have the number of columns restored otherwise the restored course will retain the
   number of columns it previously had or the default in the 'cnconfig.php' file as mentioned in the 'Installation'
   instructions above depending on the situation.
3. Note: I believe that if you restore a Columns's course on an installation that does not have the
         format then it will work and become the default course format.  However the column data will not be
         stored if you install Columns's at a later date.

Reporting Issues
================
Before reporting an issue, please ensure that you are running the latest version for your release of Moodle.  The primary
release area is located on https://moodle.org/plugins/view.php?plugin=format_columns.  It is also essential that you are
operating the required version of Moodle as stated at the top - this is because the format relies on core functionality that
is out of its control.

All 'Columns' does is integrate with the course page and control it's layout, therefore what may appear to be an issue
with the format is in fact to do with a theme or core component.  Please be confident that it is an issue with 'Columns'
but if in doubt, ask.

I operate a policy that I will fix all genuine issues for free.  Improvements are at my discretion.  I am happy to make bespoke
customisations / improvements for a negotiated fee. 

When reporting an issue you can post in the course format's forum on Moodle.org (currently 'moodle.org/mod/forum/view.php?id=47')
or contact me direct (details at the bottom).

It is essential that you provide as much information as possible, the critical information being the contents of the format's 
version.php file.  Other version information such as specific Moodle version, theme name and version also helps.  A screen shot
can be really useful in visualising the issue along with any files you consider to be relevant.

Version Information
===================
See Changes.md

References
==========
Collapsed Topics Format - Column code migrated from - https://moodle.org/plugins/view.php?plugin=format_topcoll

G J Barnard MSc. BSc(Hons)(Sndw). MBCS. CEng. CITP. PGCE.
Moodle profile: moodle.org/user/profile.php?id=442195.
Web profile   : about.me/gjbarnard

Changes.md
Version Information
===================
Version 3.0.1.2
  1. Fix horizontal columns in one section per page mode.

Version 3.0.1.1
  1. Remove 'section_right_content' method.  Prevents warning when editing.
  2. Set side widths when editing.

Version 3.0.1
  1. First version for Moodle 3.0.
  2. Added TravisCI support.
  3. Implement MDL-26226.
  4. Course format API changes for M3.0.
  5. Added 'Format responsive' setting from 'Collapsed Topics'.
  6. Improve styles for both horizontal and vertical orientations.
  7. Renamed plugin from 'Columns format' to just 'Columns'.

Version 2.9.1.1
  1. Update readme.

Version 2.9.1
  1. First version for Moodle 2.9.

16th November 2014 Version 2.8.1
  1. Stable version for Moodle 2.8.

10th November 2014 Version 2.8.0.1 - Release Candidate
  1. Release candidate for Moodle 2.8 - NOT for production servers.

20th May 2014 Version 2.7.1 - Stable.
  1. Stable release for M2.7.
  2. Fixed hidden sections break flow in Bootstrap V3 based themes.

23rd April 2014 Version 2.7.0.1 - BETA
  1. First beta version for Moodle 2.7beta.
  2. Copied 'assist layout' styles from Collapsed Topics 2.6.1.4 and 2.6.1.5

20th February 2014 Version 2.6.1.1
  1. Fixed slight coding fault in 'print_multiple_section_page' of 'renderer.php'.
  2. Refactored 'print_single_section_page' in 'renderer.php' for maintenance.
  3. Refactoring for the 'Elegance' theme: https://github.com/moodleman/moodle-theme_elegance.
  4. Fixed slight context call issue in 'lib.php'.

18th November 2013 Version 2.6.1
  1. Stable release for Moodle 2.6.

14th November 2013 Version 2.6.0.1
  1. Initial BETA code for Moodle 2.6.
  2. Changes for 'Accessibility' based upon MDL-41252.
  3. Fully implemented MDL-39542.

19th August 2013 Version 2.5.1.2
  1. Fixed issue with the 'float: left' CSS style when used to ensure that the columns were displayed correctly in the
     'vertical' column orientation.  The fix is to use 'display: inline-block' instead but this does not work in IE7, so as
     it does in IE8+ and other browsers I'm going to have to go with it.  Thanks to Ed Przyzycki for reporting this via
     Collapsed Topics.

20th June 2013 Version 2.5.1.1.
  1. Fixed issue with null '$context' in 'renderer.php' thanks to 'Jez H' for reporting this.
  2. Implemented MDL-39764 to fix maxsections < numsections issue.
  3. Added small icon which shows up when updating.
  4. Changes to 'renderer.php' because of MDL-21097.
  5. Reversed the order of the history in this file for easy reading.

14th May 2013 Version 2.5.1 - Stable
  1. First stable version for Moodle 2.5 stable.

12th May 2013 Version 2.5.0.2 - Beta
  1. Minor typo fixes.
  2. Tidied up 'format.php' to use same logic for styles as in Collapsed Topics.
  3. Changes for MDL-39542.

8th May 2013 - Version 2.5.0.1 - Beta
  1. Fixed "When in 'Show one section per page' mode and the column orientation is set to 'Horizontal' the sections on the main
     page do not fill their correct width.  This is due to the use of the 'section_summary()' method which needs to be changed
     within the format to set the calculated width on the 'li' tag." because the core fix I submitted on MDL-39099 has now
     been integrated.  Thus requiring version 2013050200.00 2.5beta+ (Build: 20130502).
  2. Removed '.jumpmenu' from styles.css because of MDL-38907.
  3. Moved 'float: left' to styles.css for Henrik Thorn - CONTRIB-4198.
  4. Improvements for MDL-34917.
  5. Realised that Tablets have more space, so allow two columns even when two or more are set.
  6. Updated core API calls to M2.5.
  7. Moved all 'cnconfig.php' default functionalty to 'Site Administration -> Plugins -> Course formats -> Columns'
     so that defaults can be changed by the administrator from within Moodle without resorting to code changes.
  8. Added capability 'format/columns:changecolumns' to editing teachers and managers such that site administrators can choose to
     disable functionality through roles if they wish.  In order for this to work the version number must be updated.
  9. Added automatic 'Purge all caches' when upgrading.  If this appears not to work by lack of display etc. then perform a
     manual 'Purge all caches' under 'Home -> Site administration -> Development -> Purge all caches'.

6th March 2013 - Version 2.4.1 - Stable.
  1. Fixed issue with strings referencing Collapsed Topics format thanks to Jago Brown.
  2. Made stable as appears to be so.

24th February 2013 - Version 2.4.0.6 - Beta - Do not install on production sites.
  1. Changes because of MDL-37976.
  2. Changes because of MDL-37901.

29th January 2013 - Version 2.4.0.5 - Beta - Do not install on production sites.
  1. Implemented the ability to set horizontal and vertical column orientation.

lib.php
<?php
// This file is part of Moodle - http://moodle.org/
//
// Moodle is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// Moodle is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.

/**
 * Columns Information
 *
 * A topic based format that allows the topics to be arranged in columns except 0.
 * Full installation instructions, code adaptions and credits are included in the 'Readme.txt' file.
 *
 * @package    course/format
 * @subpackage columns
 * @version    See the value of '$plugin->version' in version.php.
 * @copyright  &copy; 2012-onwards G J Barnard in respect to modifications of standard topics format.
 * @author     G J Barnard - gjbarnard at gmail dot com and {@link http://moodle.org/user/profile.php?id=442195}
 * @author     Based on code originally written by Dan Poltawski.
 * @license    http://www.gnu.org/copyleft/gpl.html GNU Public License
 *
 */
require_once($CFG->dirroot . '/course/format/lib.php'); // For format_base.

class format_columns extends format_base {

    private $settings;

    /**
     * Returns the format's settings and gets them if they do not exist.
     * @return type The settings as an array.
     */
    public function get_settings() {
        if (empty($this->settings) == true) {
            $this->settings = $this->get_format_options();
        }
        return $this->settings;
    }

    /**
     * Indicates this format uses sections.
     *
     * @return bool Returns true
     */
    public function uses_sections() {
        return true;
    }

    /**
     * Returns the default section name for the format.
     *
     * @param stdClass $section Section object from database or just field course_sections section
     * @return string The default value for the section name.
     */
    public function get_default_section_name($section) {
        /* Follow the same logic so that this method is supported.  The MDL-51610 enchancement refactored things,
           but that is not appropriate for us. */
        return $this->get_section_name($section);
    }

    /**
     * Gets the name for the provided section.
     *
     * @param stdClass $section The section.
     * @return string The section name.
     */
    public function get_section_name($section) {
        $course = $this->get_course();
        $section = $this->get_section($section);
        // We can't add a node without any text.
        if ((string) $section->name !== '') {
            return format_string($section->name, true, array('context' => context_course::instance($course->id)));
        } else if ($section->section == 0) {
            return get_string('section0name', 'format_columns');
        } else {
            return get_string('topic') . ' ' . $section->section;
        }
    }

    /**
     * The URL to use for the specified course (with section)
     *
     * @param int|stdClass $section Section object from database or just field course_sections.section
     *     if omitted the course view page is returned
     * @param array $options options for view URL. At the moment core uses:
     *     'navigation' (bool) if true and section has no separate page, the function returns null
     *     'sr' (int) used by multipage formats to specify to which section to return
     * @return null|moodle_url
     */
    public function get_view_url($section, $options = array()) {
        $course = $this->get_course();
        $url = new moodle_url('/course/view.php', array('id' => $course->id));

        $sr = null;
        if (array_key_exists('sr', $options)) {
            $sr = $options['sr'];
        }
        if (is_object($section)) {
            $sectionno = $section->section;
        } else {
            $sectionno = $section;
        }
        if ($sectionno !== null) {
            if ($sr !== null) {
                if ($sr) {
                    $usercoursedisplay = COURSE_DISPLAY_MULTIPAGE;
                    $sectionno = $sr;
                } else {
                    $usercoursedisplay = COURSE_DISPLAY_SINGLEPAGE;
                }
            } else {
                $usercoursedisplay = $course->coursedisplay;
            }
            if ($sectionno != 0 && $usercoursedisplay == COURSE_DISPLAY_MULTIPAGE) {
                $url->param('section', $sectionno);
            } else {
                global $CFG;
                if (empty($CFG->linkcoursesections) && !empty($options['navigation'])) {
                    return null;
                }
                $url->set_anchor('section-' . $sectionno);
            }
        }
        return $url;
    }

    /**
     * Returns the information about the ajax support in the given source format
     *
     * The returned object's property (boolean)capable indicates that
     * the course format supports Moodle course ajax features.
     * The property (array)testedbrowsers can be used as a parameter for {@link ajaxenabled()}.
     *
     * @return stdClass
     */
    public function supports_ajax() {
        $ajaxsupport = new stdClass();
        $ajaxsupport->capable = true;
        return $ajaxsupport;
    }

    /**
     * Custom action after section has been moved in AJAX mode
     *
     * Used in course/rest.php
     *
     * @return array This will be passed in ajax respose
     */
    public function ajax_section_move() {
        global $PAGE;
        $titles = array();
        $course = $this->get_course();
        $modinfo = get_fast_modinfo($course);
        $renderer = $this->get_renderer($PAGE);
        if ($renderer && ($sections = $modinfo->get_section_info_all())) {
            foreach ($sections as $number => $section) {
                $titles[$number] = $renderer->section_title($section, $course);
            }
        }
        return array('sectiontitles' => $titles, 'action' => 'move');
    }

    /**
     * Returns the list of blocks to be automatically added for the newly created course
     *
     * @return array of default blocks, must contain two keys BLOCK_POS_LEFT and BLOCK_POS_RIGHT
     *     each of values is an array of block names (for left and right side columns)
     */
    public function get_default_blocks() {
        return array(
            BLOCK_POS_LEFT => array(),
            BLOCK_POS_RIGHT => array('search_forums', 'news_items', 'calendar_upcoming', 'recent_activity')
        );
    }

    /**
     * Definitions of the additional options that this course format uses for course
     *
     * Columns format uses the following options:
     * - coursedisplay
     * - numsections
     * - hiddensections
     *
     * @param bool $foreditform
     * @return array of options
     */
    public function course_format_options($foreditform = false) {
        static $courseformatoptions = false;

        if ($courseformatoptions === false) {
            $courseconfig = get_config('moodlecourse');
            $courseformatoptions = array(
                'numsections' => array(
                    'default' => $courseconfig->numsections,
                    'type' => PARAM_INT,
                ),
                'hiddensections' => array(
                    'default' => $courseconfig->hiddensections,
                    'type' => PARAM_INT,
                ),
                'coursedisplay' => array(
                    'default' => get_config('format_columns', 'defaultcoursedisplay'),
                    'type' => PARAM_INT,
                ),
                'columns' => array(
                    'default' => get_config('format_columns', 'defaultcolumns'),
                    'type' => PARAM_INT,
                ),
                'columnorientation' => array(
                    'default' => get_config('format_columns', 'defaultcolumnorientation'),
                    'type' => PARAM_INT,
                )
            );
        }
        if ($foreditform && !isset($courseformatoptions['coursedisplay']['label'])) {
            global $COURSE;
            $coursecontext = context_course::instance($COURSE->id);

            $courseconfig = get_config('moodlecourse');
            $sectionmenu = array();
            for ($i = 0; $i <= $courseconfig->maxsections; $i++) {
                $sectionmenu[$i] = "$i";
            }
            $courseformatoptionsedit = array(
                'numsections' => array(
                    'label' => new lang_string('numbersections', 'format_columns'),
                    'element_type' => 'select',
                    'element_attributes' => array($sectionmenu),
                ),
                'hiddensections' => array(
                    'label' => new lang_string('hiddensections'),
                    'help' => 'hiddensections',
                    'help_component' => 'moodle',
                    'element_type' => 'select',
                    'element_attributes' => array(
                        array(
                            0 => new lang_string('hiddensectionscollapsed'),
                            1 => new lang_string('hiddensectionsinvisible')
                        )
                    ),
                ),
                'coursedisplay' => array(
                    'label' => new lang_string('coursedisplay'),
                    'element_type' => 'select',
                    'element_attributes' => array(
                        array(
                            COURSE_DISPLAY_SINGLEPAGE => new lang_string('coursedisplay_single'),
                            COURSE_DISPLAY_MULTIPAGE => new lang_string('coursedisplay_multi')
                        )
                    ),
                    'help' => 'coursedisplay',
                    'help_component' => 'moodle',
                )
            );
            if (has_capability('format/columns:changecolumns', $coursecontext)) {
                $courseformatoptionsedit['columns'] = array(
                    'label' => new lang_string('setcolumns', 'format_columns'),
                    'help' => 'setcolumns',
                    'help_component' => 'format_columns',
                    'element_type' => 'select',
                    'element_attributes' => array(
                        array(1 => get_string('one', 'format_columns'), // Default.
                            2 => get_string('two', 'format_columns'), // Two.
                            3 => get_string('three', 'format_columns'), // Three.
                            4 => get_string('four', 'format_columns')) // Four.
                    )
                );
                $courseformatoptionsedit['columnorientation'] = array(
                    'label' => new lang_string('setcolumnorientation', 'format_columns'),
                    'help' => 'setcolumnorientation',
                    'help_component' => 'format_columns',
                    'element_type' => 'select',
                    'element_attributes' => array(
                        array(1 => get_string('columnvertical', 'format_columns'),
                              2 => get_string('columnhorizontal', 'format_columns')) // Default.
                    )
                );
            } else {
                $courseformatoptionsedit['columns'] = array(
                    'label' => new lang_string('setcolumns', 'format_columns'), 'element_type' => 'hidden');
                $courseformatoptionsedit['columnorientation'] = array(
                    'label' => new lang_string('setcolumnorientation', 'format_columns'), 'element_type' => 'hidden');
            }
            $courseformatoptions = array_merge_recursive($courseformatoptions, $courseformatoptionsedit);
        }
        return $courseformatoptions;
    }

    /**
     * Adds format options elements to the course/section edit form
     *
     * This function is called from {@link course_edit_form::definition_after_data()}
     *
     * @param MoodleQuickForm $mform form the elements are added to
     * @param bool $forsection 'true' if this is a section edit form, 'false' if this is course edit form
     * @return array array of references to the added form elements
     */
    public function create_edit_form_elements(&$mform, $forsection = false) {

        $elements = parent::create_edit_form_elements($mform, $forsection);

        if ($forsection == false) {
            global $COURSE, $USER;
            /* Increase the number of sections combo box values if the user has increased the number of sections
               using the icon on the course page beyond course 'maxsections' or course 'maxsections' has been
               reduced below the number of sections already set for the course on the site administration course
               defaults page.  This is so that the number of sections is not reduced leaving unintended orphaned
               activities / resources. */
            $maxsections = get_config('moodlecourse', 'maxsections');
            $numsections = $mform->getElementValue('numsections');
            $numsections = $numsections[0];
            if ($numsections > $maxsections) {
                $element = $mform->getElement('numsections');
                for ($i = $maxsections + 1; $i <= $numsections; $i++) {
                    $element->addOption("$i", $i);
                }
            }

            $coursecontext = context_course::instance($COURSE->id);

            $changecolumns = has_capability('format/columns:changecolumns', $coursecontext);
            $resetall = is_siteadmin($USER); // Site admins only.

            if ($changecolumns || $resetall) {
                $elements[] = $mform->addElement('header', 'cnreset', get_string('cnreset', 'format_columns'));
            }

            if ($changecolumns) {
                $mform->addHelpButton('cnreset', 'cnreset', 'format_columns', '', true);
                $elements[] = $mform->addElement('checkbox', 'resetcolumns', get_string('resetcolumns', 'format_columns'), false);
                $mform->addHelpButton('resetcolumns', 'resetcolumns', 'format_columns', '', true);
            }

            if (is_siteadmin($USER)) {
                $elements[] = $mform->addElement('checkbox', 'resetallcolumns',
                    get_string('resetallcolumns', 'format_columns'), false);
                $mform->addHelpButton('resetallcolumns', 'resetallcolumns', 'format_columns', '', true);
            }
        }

        return $elements;
    }

    /**
     * Updates format options for a course
     *
     * In case if course format was changed to 'Columns', we try to copy options
     * 'coursedisplay', 'numsections' and 'hiddensections' from the previous format.
     * If previous course format did not have 'numsections' option, we populate it with the
     * current number of sections.  The defaults will come from 'course_format_options'.
     *
     * @param stdClass|array $data return value from {@link moodleform::get_data()} or array with data
     * @param stdClass $oldcourse if this function is called from {@link update_course()}
     *     this object contains information about the course before update
     * @return bool whether there were any changes to the options values
     */
    public function update_course_format_options($data, $oldcourse = null) {
        global $DB; // MDL-37976.
        /* Notes: Using 'unset' to really ensure that the reset form elements never get into the database.
                  This has to be done here so that the reset occurs after we have done updates such that the
                  reset itself is not seen as an update. */
        $resetcolumns = false;
        $resetallcolumns = false;
        if (isset($data->resetcolumns) == true) {
            $resetcolumns = true;
            unset($data->resetcolumns);
        }
        if (isset($data->resetallcolumns) == true) {
            $resetallcolumns = true;
            unset($data->resetallcolumns);
        }

        if ($oldcourse !== null) {
            $data = (array) $data;
            $oldcourse = (array) $oldcourse;
            $options = $this->course_format_options();
            foreach ($options as $key => $unused) {
                if (!array_key_exists($key, $data)) {
                    if (array_key_exists($key, $oldcourse)) {
                        $data[$key] = $oldcourse[$key];
                    } else if ($key === 'numsections') {
                        /* If previous format does not have the field 'numsections'
                           and $data['numsections'] is not set,
                           we fill it with the maximum section number from the DB. */
                        $maxsection = $DB->get_field_sql('SELECT max(section) from {course_sections}
                            WHERE course = ?', array($this->courseid));
                        if ($maxsection) {
                            // If there are no sections, or just default 0-section, 'numsections' will be set to default.
                            $data['numsections'] = $maxsection;
                        }
                    }
                }
            }
        }
        $changes = $this->update_format_options($data);

        // Now we can do the reset.
        if ($resetcolumns == true) {
            $this->reset_columns_setting($this->courseid);
            $changes = true;
        }

        if ($resetallcolumns == true) {
            $this->reset_columns_setting(0);
            $changes = true;
        }

        return $changes;
    }

    /**
     * Resets the format setting to the default.
     * @param int $courseid If not 0, then a specific course to reset.
     */
    public function reset_columns_setting($courseid) {
        global $DB, $USER, $COURSE;

        $currentcourseid = 0;
        if ($courseid == 0) {
            $records = $DB->get_records('course_format_options', array('format' => $this->format), '', 'id, courseid');
        } else {
            $records = $DB->get_records('course_format_options', array(
                'courseid' => $courseid, 'format' => $this->format), '', 'id, courseid');
        }

        $resetallifall = ((is_siteadmin($USER)) || ($courseid != 0)); // Will be true if reset all capability or a single course.

        $coursecontext = context_course::instance($COURSE->id);

        $updatedata = array();
        if (has_capability('format/columns:changecolumns', $coursecontext) && $resetallifall) {
            $updatedata['coursedisplay'] = get_config('format_columns', 'defaultcoursedisplay');
            $updatedata['columns'] = get_config('format_columns', 'defaultcolumns');
            $updatedata['columnorientation'] = get_config('format_columns', 'defaultcolumnorientation');
        }

        foreach ($records as $record) {
            if ($currentcourseid != $record->courseid) {
                $currentcourseid = $record->courseid; // Only do once per course.
                $ourcourseid = $this->courseid;
                $this->courseid = $currentcourseid;
                $this->update_format_options($updatedata);
                $this->courseid = $ourcourseid;
            }
        }
    }

    /**
     * Restores the course settings when restoring a Moodle 2.3 or below (bar 1.9) course and sets the settings when upgrading
     * from a prevous version.  Hence no need for 'coursedisplay' as that is a core rather than CN specific setting and not
     * in the old 'format_columns_settings' table.
     * @param int $courseid If not 0, then a specific course to reset.
     * @param int $columns The columns to use, see cnconfig.php.
     */
    public function restore_columns_setting($courseid, $columns) {
        $currentcourseid = $this->courseid;  // Save for later - stack data model.
        $this->courseid = $courseid;
        // Create data array.
        $data = array('columns' => $columns);

        $this->update_course_format_options($data);

        $this->courseid = $currentcourseid;
    }

    /**
     * Updates the number of columns when the renderer detects that they are wrong.
     * @param int $columns The columns to use, see cnconfig.php.
     */
    public function update_columns_columns_setting($columns) {
        // Create data array.
        $data = array('columns' => $columns);

        $this->update_course_format_options($data);
    }
}

/**
 * Used to display the course structure for a course where format=Columns
 *
 * This is called automatically by {@link load_course()} if the current course
 * format = Columns.
 *
 * @param navigation_node $navigation The course node.
 * @param array $path An array of keys to the course node.
 * @param stdClass $course The course we are loading the section for.
 */
function callback_columns_load_content(&$navigation, $course, $coursenode) {
    return $navigation->load_generic_course_sections($course, $coursenode, 'columns');
}

/**
 * The string that is used to describe a section of the course.
 *
 * @return string The section description.
 */
function callback_columns_definition() {
    return get_string('sectionname', 'format_columns');
}
